java.lang.IllegalStateException: Timer already cancelled.

I have a project that has a lot of changes and `git pull` takes a long time. When I switch to another project (eg. https://github.com/quarkusio/quarkus) the following error is displayed in the error logs. I don't think it's easily reproducible, but I thought it would be nice to report: 


```
   Exception at Workspace Model event handling
   
   java.lang.RuntimeException: Cannot invoke (class=, method=changed, topic=WorkspaceModelChangeListener)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:676)
	at com.intellij.util.messages.impl.MessageBusImplKt.executeOrAddToQueue(MessageBusImpl.kt:491)
	at com.intellij.util.messages.impl.MessagePublisher.publish$intellij_platform_core(MessageBusImpl.kt:471)
	at com.intellij.util.messages.impl.MessagePublisher.invoke(MessageBusImpl.kt:448)
	at jdk.proxy2/jdk.proxy2.$Proxy79.changed(Unknown Source)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl$onChanged$3.invoke(WorkspaceModelImpl.kt:347)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl$onChanged$3.invoke(WorkspaceModelImpl.kt:346)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl.logErrorOnEventHandling(WorkspaceModelImpl.kt:386)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl.onChanged(WorkspaceModelImpl.kt:346)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl.access$onChanged(WorkspaceModelImpl.kt:36)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl$replaceProjectModel$1$2.invoke(WorkspaceModelImpl.kt:299)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl$replaceProjectModel$1$2.invoke(WorkspaceModelImpl.kt:299)
	at com.intellij.platform.workspace.storage.impl.VersionedEntityStorageImpl.replace(VersionedEntityStorageImpl.kt:190)
	at com.intellij.workspaceModel.ide.impl.WorkspaceModelImpl.replaceProjectModel(WorkspaceModelImpl.kt:299)
	at org.jetbrains.idea.maven.importing.workspaceModel.WorkspaceProjectImporter$Companion.updateProjectModelFastOrSlow$lambda$3(WorkspaceProjectImporter.kt:622)
	at org.jetbrains.idea.maven.utils.MavenUtil.runAndFinishTestRunnable(MavenUtil.java:187)
	at org.jetbrains.idea.maven.utils.MavenUtil.lambda$invokeAndWaitWriteAction$2(MavenUtil.java:243)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:992)
	at org.jetbrains.idea.maven.utils.MavenUtil.lambda$invokeAndWaitWriteAction$3(MavenUtil.java:243)
	at com.intellij.util.DisposeAwareRunnable.run(DisposeAwareRunnable.java:69)
	at com.intellij.util.concurrency.ContextRunnable.run(ContextRunnable.java:24)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:208)
	at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:190)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:861)
	at com.intellij.openapi.application.impl.ApplicationImpl$4.run(ApplicationImpl.java:478)
	at com.intellij.openapi.application.impl.LaterInvocator$1.run(LaterInvocator.java:100)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:79)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:121)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:41)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:792)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:739)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:733)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:761)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:685)
	at com.intellij.ide.IdeEventQueue._dispatchEvent$lambda$10(IdeEventQueue.kt:589)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWithoutImplicitRead(ApplicationImpl.java:1485)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:589)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:67)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:369)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:368)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:787)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:368)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:363)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:992)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:105)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:992)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$7(IdeEventQueue.kt:363)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:861)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:405)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.lang.IllegalStateException: Timer already cancelled.
	at java.base/java.util.Timer.sched(Timer.java:409)
	at java.base/java.util.Timer.schedule(Timer.java:205)
	at com.redhat.devtools.intellij.lsp4mp4ij.classpath.ClasspathResourceChangedNotifier.asyncNotifyChanges(ClasspathResourceChangedNotifier.java:95)
	at com.redhat.devtools.intellij.lsp4mp4ij.classpath.ClasspathResourceChangedNotifier.addLibrary(ClasspathResourceChangedNotifier.java:63)
	at com.redhat.devtools.intellij.lsp4mp4ij.classpath.ClasspathResourceChangedListener.handleLibraryUpdate(ClasspathResourceChangedListener.java:73)
	at com.redhat.devtools.intellij.lsp4mp4ij.classpath.ClasspathResourceChangedListener.afterLibraryRemoved(ClasspathResourceChangedListener.java:66)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.intellij.util.EventDispatcher.dispatchVoidMethod(EventDispatcher.java:118)
	at com.intellij.util.EventDispatcher.lambda$createMulticaster$1(EventDispatcher.java:83)
	at jdk.proxy2/jdk.proxy2.$Proxy81.afterLibraryRemoved(Unknown Source)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.library.ProjectLibraryTableBridgeImpl$1.changed(ProjectLibraryTableBridgeImpl.kt:106)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeMethod(MessageBusImpl.kt:699)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:659)
	... 58 more

```

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerWrapper.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerWrapper.java
@@ -497,7 +497,12 @@ public class LanguageServerWrapper implements Disposable {
         timer.schedule(new TimerTask() {
             @Override
             public void run() {
-                stop();
+                try {
+                    stop();
+                } catch (Throwable t) {
+                    //Need to catch time task exceptions, or it will cancel the timer
+                    LOGGER.error("Failed to stop language server "+LanguageServerWrapper.this.serverDefinition.id, t);
+                }
             }
         }, TimeUnit.SECONDS.toMillis(this.serverDefinition.lastDocumentDisconnectedTimeout));
     }

--- a/src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/classpath/ClasspathResourceChangedNotifier.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/classpath/ClasspathResourceChangedNotifier.java
@@ -24,6 +24,8 @@ import com.intellij.openapi.roots.libraries.Library;
 import com.intellij.openapi.util.Pair;
 import com.intellij.openapi.vfs.VirtualFile;
 import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.HashSet;
 import java.util.Set;
@@ -34,6 +36,8 @@ import java.util.TimerTask;
  * Source file change notifier with a debounce mode.
  */
 public class ClasspathResourceChangedNotifier implements Disposable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClasspathResourceChangedNotifier.class);
 
     private static final long DEBOUNCE_DELAY = 1000;
 
@@ -80,7 +84,12 @@ public class ClasspathResourceChangedNotifier implements Disposable {
             debounceTask = new TimerTask() {
                 @Override
                 public void run() {
-                    notifyChanges();
+                    try {
+                        notifyChanges();
+                    } catch (Throwable t) {
+                        //Need to catch time task exceptions, or it will cancel the timer
+                        LOGGER.error("Failed to notify classpath resource change", t);
+                    }
                 }
             };
 

